### PR TITLE
Add M118 command to virtual printer

### PR DESCRIPTION
**Pull request by max246**
* [ x] Your changes are not possible to do through a plugin and relevant
    to a large audience (ideally all users of OctoPrint)
  * [x ] If your changes are large or otherwise disruptive: You have
    made sure your changes don't interfere with current development by
    talking it through with the maintainers, e.g. through a
    Brainstorming ticket
  * [x ] Your PR targets OctoPrint's devel branch if it's a completely
    new feature, or maintenance if it's a bug fix or improvement of
    existing functionality for the current stable version (no PRs
    against master or anything else please)
  * [ x] Your PR was opened from a custom branch on your repository
    (no PRs from your version of master, maintenance or devel please),
    e.g. dev/my_new_feature or fix/my_bugfix
  * [x ] Your PR only contains relevant changes: no unrelated files,
    no dead code, ideally only one commit - rebase and squash your PR
    if necessary!
  * [x] Your changes follow the existing coding style
  * [ x] If your changes include style sheets: You have modified the
    .less source files, not the .css files (those are generated with
    lessc)
  * [x ] You have tested your changes (please state how!) - ideally you
    have added unit tests
  * [x ] You have run the existing unit tests against your changes and
    nothing broke
  * [x ] You have added yourself to the AUTHORS.md file :)

#### What does this PR do and why is it necessary?

Implemented M118 command in the virtual printer which was requested from ticket #2191 

#### How was it tested? How can it be tested by the reviewer?

Run the software locally and run the virtual printer to observe the functionality of the command

#### Any background context you want to provide?

 No 

#### What are the relevant tickets if any?

Ticket #2191 

#### Further notes

I have updated the method of formatting string for M117 which is the preferred method from the author of this project

### CODE REVIEW

**Review by cp2004 — CHANGES_REQUESTED** [review]
Can the command support the A/E parameters? https://marlinfw.org/docs/gcode/M118.html
I guess it is not in the reprap spec page, and they don't actually do much other than prefixing the response.

**Comment by foosel** [comment]
@max246 Have you had a chance to look at the review comments by @cp2004?

**Comment by max246** [comment]
> @max246 Have you had a chance to look at the review comments by @cp2004?

Sorry I didnt, I was off for a bit, I should pick it up again and see how were done :) might be more active this month

File: src/octoprint/plugins/virtual_printer/virtual.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 import collections
 import json
+import logging
```
[the remark this answers is not in the document]
Comment (max246) [reply src/octoprint/plugins/virtual_printer/virtual.py]: Done it

File: src/octoprint/plugins/virtual_printer/virtual.py [hunk 2]
```diff
@@ -697,10 +698,18 @@ def _gcode_M117(self, data: str) -> None:
...
+                self._send(f"echo:{result}")
             except AttributeError:
                 self._send("echo:")
 
+    def _gcode_M118(self, data: str) -> None:
+        try:
+            result = re.search(r"M118\s+(.*)", data).group(1)
+            self._send(f"{result}")
+        except Exception:
+            logging.debug("Error sending back echo command M118")
```
[the remark this answers is not in the document]
Comment (max246) [reply src/octoprint/plugins/virtual_printer/virtual.py]: Done it

**Comment by max246** [comment]
> 

Implemented the two options and unknown command

File: src/octoprint/plugins/virtual_printer/virtual.py [hunk 3]
```diff
@@ -697,10 +697,24 @@ def _gcode_M117(self, data: str) -> None:
...
+            parameter, result = re.search(r"M118 (\w{2}) (.*)", data).groups()
+            if "A1" in parameter.upper():
+                self._send(f"{result}")
+            elif "E1" in parameter.upper():
+                self._send(f"echo:{result}")
+            else:
+                self._logger.exception(f"{parameter} is not recognised")
+                raise Exception
+        except Exception:
+            self._logger.debug("Error sending back echo command M118")
```
Comment (foosel) [inline src/octoprint/plugins/virtual_printer/virtual.py:716]: Either I misunderstand the docs, or the logic here is wrong.

First of all, `A1` and `E1` are optional, they do not have to be there. `M118 hello` should still work, it doesn't here. Then, `A1` is supposed to prefix the output with a `//`, it doesn't here.

How about this instead, this also adheres more strictly to the actual firmware behaviour and doesn't try to capture any two characters as parameter, but only those that are defined:

```suggestion
        match = re.search(r"M118 (?:(?P<parameter>A1|E1|Pn[012])\s)?(?P<text>.*)")
        if not match:
            self._send("Unrecognized command parameters for M118")
        
        text = match.groupdict().get("text")
        parameter = match.groupdict().get("parameter")
        
        if parameter == "A1":
            self._send(f"//{text}")
        elif parameter == "E1":
            self._send(f"echo:{text}")
        else:
            self._send(text)

```

See [here](https://regex101.com/r/tVDSjP/1) for some testing of that regex.

Cross checking this with [the firmware implementation](https://github.com/MarlinFirmware/Marlin/blob/42d30ed102308f0c5cb04b70dc55f9f37012a157/Marlin/src/gcode/host/M118.cpp#L35) looks like it matches in behaviour.

**Review by foosel — CHANGES_REQUESTED** [review]

File: src/octoprint/plugins/virtual_printer/virtual.py (hunk 3 above)
Comment (max246) [reply src/octoprint/plugins/virtual_printer/virtual.py:716]: that sounds good

**Review by cp2004 — APPROVED** [review]
Haven't tested but the logic does look good to me now.

**Review by foosel — APPROVED** [review]

**Pull request merged**